Prepare AMP support for imx6xx

dbaluta opened the pull request:
imxrt6xx are dual core devices featuring an ARM Cortex-M33 core and an Cadence Xtensa HIFI4 Audio DSP.

This patchseries basically re-organize m33 existing support so that it allows adding support for HIFI4 DSP in a next Pull Request.

dbaluta committed:
soc: nxp: imxrt: Prepare imxrt6xx soc for AMP support

imxrt6xx are dual core devices featuring an ARM Cortex-M33
core and an Cadence Xtensa HIFI4 Audio DSP.

Currently only m33 core is supported. In order to support
the Cadence DSP we need first to do some code-reorganization
for m33.

We start by moving all cm33 related code to its own directory
and introduce the cpuclusters property in soc.yml file.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

dbaluta committed:
boards: nxp: Rename mimxrt685_evk to mimxrt685_evk_mimxrt685s_cm33

In preparation for adding AMP support for i.MX RT6xx family we need
to rename existing cm33 support files to more specific names.

e.g mimxrt685_evk.dts -> mimxrt685_evk_mimxrt685s_cm33.dts

This will allow us to later add support for Cadence DSP found on i.MX
RT6xx series.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

danieldegrasse commented on boards/nxp/mimxrt685_evk/mimxrt685_evk_mimxrt685s_cm33-pinctrl.dtsi:
Is there a reason not to share the pinctrl file between cores? I personally prefer that approach, since it allows us to spot conflicts between each core's pin selections a bit easier.

danieldegrasse replied:
Sure, that is totally fair- I think I was probably just doing bulk renames when I made the RT595 port. I honestly only realized the value when looking at the iMX boards (since the RT595 DSP support doesn't enable peripherals like the USART)

danieldegrasse reviewed (commented):
I believe the twister failures you are seeing are because the board identifier needs to be updated in the board YAML, as well as in any twister tests that currently reference it

dbaluta commented:
> I believe the twister failures you are seeing are because the board identifier needs to be updated in the board YAML, as well as in any twister tests that currently reference it

Yes, just pushed the changes to update the twister tests! Should be fine now.

dbaluta commented:
Changes since v2:
- no longer rename pinctrl file in order to share it between cores
- updated board identifier in twister tests

dbaluta commented:
Changes since v3:
- update identifier in board yaml file also

LaurentiuM1234 reviewed (commented):
Looks ok. Just one question: why move `pinctrl_soc.h` to the `cm33` folder? Is this supposed to be cluster-dependent? Doesn't look like it from https://github.com/zephyrproject-rtos/zephyr/pull/65229.

dbaluta commented:
> Looks ok. Just one question: why move `pinctrl_soc.h` to the `cm33` folder? Is this supposed to be cluster-dependent? Doesn't look like it from #65229.

Yes, this can be done. 

The practice was to have a separate pinctrl_soc.h but I guess that's because there were different people working on it.

``
./soc/nxp/imx/imx8m/adsp/pinctrl_soc.h
./soc/nxp/imx/imx8m/m7/pinctrl_soc.h
./soc/nxp/imx/imx8m/m4_quad/pinctrl_soc.h
./soc/nxp/imx/imx8m/m4_mini/pinctrl_soc.h
```

```
--- ./soc/nxp/imx/imx8m/m7/pinctrl_soc.h        2024-03-28 17:23:47.440062633 +0200
+++ ./soc/nxp/imx/imx8m/adsp/pinctrl_soc.h      2024-03-28 17:23:47.440062633 +0200
@@ -1,11 +1,11 @@
 /*
- * Copyright (c) 2022, NXP
+ * Copyright (c) 2023, NXP
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#ifndef ZEPHYR_SOC_ARM_NXP_MIMX8ML8_M7_PINCTRL_SOC_H_
-#define ZEPHYR_SOC_ARM_NXP_MIMX8ML8_M7_PINCTRL_SOC_H_
+#ifndef ZEPHYR_SOC_XTENSA_NXP_IMX8M_PINCTRL_SOC_H_
+#define ZEPHYR_SOC_XTENSA_NXP_IMX8M_PINCTRL_SOC_H_
 
 #include <zephyr/devicetree.h>
 #include <zephyr/types.h>
@@ -80,4 +80,4 @@
 }
 #endif
 
-#endif /* ZEPHYR_SOC_ARM_NXP_MIMX8ML8_M7_PINCTRL_SOC_H_ */
+#endif /* ZEPHYR_SOC_XTENSA_NXP_IMX8M_PINCTRL_SOC_H_ */
```

dbaluta commented:
@LaurentiuM1234 I think you are right we are using the same soc_pinctrl.h file for DSP also so I quit on moving it to cm33/ folder.

Changes since v4:
* keep soc_pinctrl.h location as it is.

danieldegrasse commented:
> I wonder what is the real reason? Maybe because different people worked on different architectures?

Almost for sure- both cores use the same pin controller, so ideally they should be sharing a driver

LaurentiuM1234 reviewed (approved):
Well, `imx8m` is somewhat of an exception. This contains more than 1 SoC.

decsny commented:
's' should be a board variant, rather than a different SOC, right?

dbaluta commented:
> 's' should be a board variant, rather than a different SOC, right?

I'm afraid I don't understand the question. Are we talking about `mimxrt685s` name or about `imx8m` example i provided above?

decsny commented:
> > 's' should be a board variant, rather than a different SOC, right?
> 
> I'm afraid I don't understand the question. Are we talking about `mimxrt685s` name or about `imx8m` example i provided above?

I meant the mimxrt685s, I am assuming the 's' means secure trustzone of the m33? which we usually do as  a board variant

danieldegrasse commented:
> @danieldegrasse can you comment on this ? I borrowed the naming from mimx595s

I honestly have no idea- as far as I know, all RT6xx part numbers have the "S" suffix on them. Since the "S" is part of the part number I would keep it. Usually when building for trustzone enabled boards, we build Zephyr in secure mode by default and provide a `ns` variant which runs with TFM, see the LPCXpresso55S69 variants:

- `lpcxpresso55s69//cpu0` (cpu0, secure)
- `lpcxpresso55s69//cpu1` (cpu1, nonsecure)
- `lpcxpresso55s69//ns` (ns , nonsecure with TFM)

decsny commented:
> > @danieldegrasse can you comment on this ? I borrowed the naming from mimx595s
> 
> I honestly have no idea- as far as I know, all RT6xx part numbers have the "S" suffix on them. Since the "S" is part of the part number I would keep it. Usually when building for trustzone enabled boards, we build Zephyr in secure mode by default and provide a `ns` variant which runs with TFM, see the LPCXpresso55S69 variants:
> 
> * `lpcxpresso55s69//cpu0` (cpu0, secure)
> * `lpcxpresso55s69//cpu1` (cpu1, nonsecure)
> * `lpcxpresso55s69//ns` (ns , nonsecure with TFM)

so I guess now the target in the future would be like `mimxrt685_evk/mimxrt685s/cm33/ns` for example for a nonsecure variant? seems confusing, unless the s means something else?

decsny commented:
update: I am looking at this buy page: https://www.nxp.com/products/processors-and-microcontrollers/arm-microcontrollers/i-mx-rt-crossover-mcus/i-mx-rt600-crossover-mcu-with-arm-cortex-m33-and-dsp-cores:i.MX-RT600?tab=Buy_Parametric_Tab#/ and I guess the S is part of the part number? so maybe nevermind, I guess I made a wrong assumption, but it still seems strange to include the S in the device name on the part of the SDK since almost every other device ends with the number, especially unfortunate given that this part is an m33 with trustzone, but oh well.
